src/ERC6909.sol

```diff
@@ -87,14 +117,109 @@ abstract contract ERC6909 is IERC6909Claims {
     //////////////////////////////////////////////////////////////*/
 
     function _mint(address receiver, uint256 id, uint256 amount) internal virtual {
-        balanceOf[receiver][id] += amount;
+        _increaseBalanceOf(receiver, id, amount);
 
         emit Transfer(msg.sender, address(0), receiver, id, amount);
     }
 
     function _burn(address sender, uint256 id, uint256 amount) internal virtual {
-        balanceOf[sender][id] -= amount;
+        _decreaseBalanceOf(sender, id, amount);
 
         emit Transfer(msg.sender, sender, address(0), id, amount);
     }
+
+    /*//////////////////////////////////////////////////////////////
+                        INTERNAL STORAGE LOGIC
+    //////////////////////////////////////////////////////////////*/
+    function _setAllowance(address owner, address spender, uint256 id, uint256 value) internal {
+        bytes32 allowanceSlot = _getAllowanceSlot(owner, spender, id);
+        /// @solidity memory-safe-assembly
+        assembly {
+            sstore(allowanceSlot, value)
+        }
+    }
+
+    function _decreaseBalanceOf(address owner, uint256 id, uint256 value) internal {
+        bytes32 balanceSlot = _getBalanceSlot(owner, id);
+        uint256 balanceOfValue;
+        /// @solidity memory-safe-assembly
+        assembly {
+            balanceOfValue := sload(balanceSlot)
+        }
+
+        balanceOfValue -= value;
+
+        /// @solidity memory-safe-assembly
+        assembly {
+            sstore(balanceSlot, balanceOfValue)
+        }
+    }
+
+    function _increaseBalanceOf(address owner, uint256 id, uint256 value) internal {
+        bytes32 balanceSlot = _getBalanceSlot(owner, id);
+        uint256 balanceOfValue;
+        /// @solidity memory-safe-assembly
+        assembly {
+            balanceOfValue := sload(balanceSlot)
+        }
+
+        balanceOfValue += value;
+
+        /// @solidity memory-safe-assembly
+        assembly {
+            sstore(balanceSlot, balanceOfValue)
+        }
+    }
+
+    function _getOperatorSlot(address owner, address spender) internal pure returns (bytes32 operatorSlot) {
+        // balanceSlot = keccak256(abi.encode(owner, spender, _operators.slot));
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x20, spender)
+            mstore(0x00, owner)
+            let pointer := mload(0x40)
+            mstore(0x40, _operators.slot)
+
+            operatorSlot := keccak256(0x00, 0x60)
+
+            mstore(0x40, pointer)
+        }
+    }
+
+    function _getBalanceSlot(address owner, uint256 id) internal pure returns (bytes32 balanceSlot) {
+        // balanceSlot = keccak256(abi.encode(owner, id, _balances.slot));
+        /// @solidity memory-safe-assembly
+        assembly {
+            mstore(0x20, id)
+            mstore(0x00, owner)
+            let pointer := mload(0x40)
+            mstore(0x40, _balances.slot)
```

Comment:
I feel like we dont need to use `_balances.slot` at all. We need something to make sure that there cannot be clashes - but i feel like we could use something just packed into the top of the owner slot?

Like save some constant 
```
bytes1 constant BALANCE_SLOT = bytes1(bytes32(keccak256("BalanceSlot")))
```

And then for the hash just put that byte into the top of the same slot as `owner`?

Reply:
that will then save going into the 0x40 slot, and no need to backup and restore the fmp too

Reply:
Same for the others too of course
And really i would put the hardcoded value in the constant and just say the calculation in a comment because constants are evaluated at runtime time 🙏

Reply:
Nice idea! Will implement it

about constants: I thought that it was fixed in one of the previous Solidity versions, isn’t it?

Reply:
Sadly I dont think so. See PR #513 where I realised it wasnt haha.